Encoding explicit `none` default produces C error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.9 df51e84.d9f5112
OS: macos, macOS, 12.7.6, 21H1320
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-6920HQ CPU @ 2.90GHz

getwd: /Users/elogir/Documents/vug
vexe: /Users/elogir/Documents/v/v
vexe mtime: 2024-12-30 14:43:11

vroot: OK, value: /Users/elogir/Documents/v
VMODULES: OK, value: /Users/elogir/.vmodules
VTMP: OK, value: /tmp/v_502

Git version: git version 2.37.1 (Apple Git-137.1)
Git vroot status: weekly.2024.52-30-gd9f51120
.git/config present: true

CC version: Apple clang version 14.0.0 (clang-1400.0.29.202)
emcc version: N/A
thirdparty/tcc status: thirdparty-macos-amd64 975f1ad8
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg src/main.v`

```v
module main

import json

struct Test {
	id ?string = none
}

fn main() {
	json.encode(Test{})
}
```

**What did you expect to see?**

Compile properly and/or warning about redundant default

**What did you see instead?**

```
src/main.v:6:15: warning: unnecessary default value of `none`: struct fields are zeroed by default
    4 | 
    5 | struct Test {
    6 |     id ?string = none
      |                  ~~~~
    7 | }
    8 |
================== C compilation error (from cc): ==============
cc: /tmp/v_502/main.01JGC0BEYNN2ZFPG22VQBJA5QX.tmp.c:1895:38: error: initializing 'byte' (aka 'unsigned char') with an expression of incompatible type 'IError' (aka 'struct IError')
cc:         main__Test res = (main__Test){.id = _const_none__,};
cc:                                             ^~~~~~~~~~~~~
cc: /tmp/v_502/main.01JGC0BEYNN2ZFPG22VQBJA5QX.tmp.c:1945:10: error: assigning to '_option_string' (aka 'struct _option_string') from incompatible type 'IError' (aka 'struct IError')
cc:                 res.id = _const_none__;
cc:                        ^ ~~~~~~~~~~~~~
cc: /tmp/v_502/main.01JGC0BEYNN2ZFPG22VQBJA5QX.tmp.c:6906:2: warning: expression result unused [-Wunused-value]
cc: (_t2);
cc:  ^~~
cc: 1 warning and 2 errors generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzcyYjI3NmYxNDRmNTg1YWU1MTZjOWYiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.YP-71xMas84yAiosVZ7sCGjUrZxfoKlHa_HFa9ifC78">Huly&reg;: <b>V_0.6-21750</b></a></sub>

--- a/vlib/json/tests/json_encode_default_option_none_test.v
+++ b/vlib/json/tests/json_encode_default_option_none_test.v
@@ -1,0 +1,11 @@
+module main
+
+import json
+
+struct Test {
+	id ?string = none
+}
+
+fn test_main() {
+	assert json.encode(Test{}) == '{}'
+}

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1236,6 +1236,14 @@ fn (mut g Gen) expr_string(expr ast.Expr) string {
 	return g.out.cut_to(pos).trim_space()
 }
 
+fn (mut g Gen) expr_string_opt(typ ast.Type, expr ast.Expr) string {
+	expr_str := g.expr_string(expr)
+	if expr is ast.None {
+		return '(${g.styp(typ)}){.state=2, .err=${expr_str}, .data={EMPTY_STRUCT_INITIALIZATION}}'
+	}
+	return expr_str
+}
+
 fn (mut g Gen) expr_string_with_cast(expr ast.Expr, typ ast.Type, exp ast.Type) string {
 	pos := g.out.len
 	// pos2 := 	g.out_parallel[g.out_idx].len
@@ -7034,7 +7042,7 @@ fn (mut g Gen) type_default_impl(typ_ ast.Type, decode_sumtype bool) string {
 									}
 								}
 							} else {
-								default_str := g.expr_string(field.default_expr)
+								default_str := g.expr_string_opt(field.typ, field.default_expr)
 								if default_str.count('\n') > 1 {
 									g.type_default_vars.writeln(default_str.all_before_last('\n'))
 									expr_str = default_str.all_after_last('\n')

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -721,7 +721,8 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 				dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${dec_name}(jsonroot_${tmp});')
 				if field.has_default_expr {
 					dec.writeln('\t} else {')
-					dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string(field.default_expr)};')
+					dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string_opt(field.typ,
+						field.default_expr)};')
 				}
 				dec.writeln('\t}')
 			} else if field_sym.kind == .enum {
@@ -755,7 +756,8 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 				}
 				if field.has_default_expr {
 					dec.writeln('\t} else {')
-					dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string(field.default_expr)};')
+					dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string_opt(field.typ,
+						field.default_expr)};')
 				}
 				dec.writeln('\t}')
 			} else if field_sym.name == 'time.Time' {
@@ -767,7 +769,8 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 				dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = time__unix(json__decode_u64(jsonroot_${tmp}));')
 				if field.has_default_expr {
 					dec.writeln('\t} else {')
-					dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string(field.default_expr)};')
+					dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string_opt(field.typ,
+						field.default_expr)};')
 				}
 				dec.writeln('\t}')
 			} else if field_sym.kind == .alias {
@@ -788,7 +791,8 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 					dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${parent_dec_name} (jsonroot_${tmp});')
 					if field.has_default_expr {
 						dec.writeln('\t} else {')
-						dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string(field.default_expr)};')
+						dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string_opt(field.typ,
+							field.default_expr)};')
 					}
 					dec.writeln('\t}')
 				} else {
@@ -799,7 +803,8 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 					dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = *(${field_type}*) ${tmp}.data;')
 					if field.has_default_expr {
 						dec.writeln('\t} else {')
-						dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string(field.default_expr)};')
+						dec.writeln('\t\t${prefix}${op}${c_name(field.name)} = ${g.expr_string_opt(field.typ,
+							field.default_expr)};')
 					}
 					dec.writeln('\t}')
 				}
@@ -838,7 +843,7 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 				}
 				if field.has_default_expr {
 					dec.writeln('\t} else {')
-					default_str := g.expr_string(field.default_expr)
+					default_str := g.expr_string_opt(field.typ, field.default_expr)
 					lines := default_str.count('\n')
 					if lines > 1 {
 						dec.writeln(default_str.all_before_last('\n'))

